utils/src/main/kotlin/com/exactpro/th2/lwdataprovider/MessageSearcher.kt

```diff
@@ -0,0 +1,195 @@
+            generateSequence {
+                if (to.isBefore(end)) {
+                    return@generateSequence null
+                }
+
+                from = to
+                to = from.minus(searchStep)
+
+                service.searchMessageGroups(
+                    requestBuilder.apply {
+                        startTimestamp = from.toTimestamp()
+                        endTimestamp = to.toTimestamp()
+                    }.build().also {
+                        K_LOGGER.debug { "Request from: $from, to: $to" }
+                    }
+                )
+            }.flatMap(Iterator<MessageSearchResponse>::asSequence)
+                .onEach { K_LOGGER.trace { "Message: ${it.toJson()}" } }
+                .firstOrNull(filter::apply)
+        }
+    }
+
+    @JvmOverloads
+    fun with(
+        requestInterval: Duration = DEFAULT_SEARCH_STEP,
+        defaultBook: String = "",
+        defaultSessionGroup: String = "",
+        defaultMessageStreams: Set<MessageStream> = emptySet(),
+    ) = MessageSearcher(service, requestInterval, defaultBook, defaultSessionGroup, defaultMessageStreams)
+
+    fun withRequestInterval(requestInterval: Duration) =
+        with(requestInterval, defaultBook, defaultSessionGroup, defaultMessageStreams)
+
+    fun withBook(defaultBook: String) =
+        with(searchStep, defaultBook, defaultSessionGroup, defaultMessageStreams)
+
+    fun withSessionGroup(defaultSessionGroup: String) =
+        with(searchStep, defaultBook, defaultSessionGroup, defaultMessageStreams)
+
+    fun withMessageStreams(defaultMessageStreams: Set<MessageStream>) =
+        with(searchStep, defaultBook, defaultSessionGroup, defaultMessageStreams)
+
+    private fun createRequestBuilder(
+        book: String,
+        sessionGroup: String,
+        messageStreams: Set<MessageStream>,
+    ) = MessageGroupsSearchRequest.newBuilder().apply {
+        bookIdBuilder.name = book
+        addMessageGroup(MessageGroupsSearchRequest.Group.newBuilder().setName(sessionGroup))
+        addAllStream(messageStreams)
+        addResponseFormats(BASE64_FORMAT)
+        searchDirection = TimeRelation.PREVIOUS
+    }
+
+    companion object {
+        private val K_LOGGER = KotlinLogging.logger { }
+        private const val BASE64_FORMAT = "BASE_64"
+
+        @JvmField
+        val DEFAULT_SEARCH_STEP: Duration = Duration.ofDays(1)
+
+        @JvmOverloads
+        fun create(
+            service: DataProviderService,
+            searchStep: Duration = DEFAULT_SEARCH_STEP,
+            defaultBook: String = "",
+            defaultSessionGroup: String = "",
+            defaultMessageStreams: Set<MessageStream> = emptySet(),
+        ) = MessageSearcher(
+            service,
+            searchStep,
+            defaultBook,
+            defaultSessionGroup,
+            defaultMessageStreams,
+        )
+
+        fun create(
+            sessionAlias: String,
+            direction: Direction,
+        ): MessageStream {
```

Comment:
We have 2 functions named `create` in one class that return different values. Probably, this might be unclear what this function returns when you read the code that uses it.
I would suggest renaming the last one to `createMessageStream`. In this case, the call place will look like:
```kotlin
val searcher = MessageSearcher.create(/*...*/)
val messageStream = MessageSearcher.createMessageStream(/*...*/)
```
Also, please add a `@JvmStatic` annotation to those methods. Otherwise, access to those methods will be available only through the `Companion` object in Java